Allow new attributes to be defined in #update

Leooo opened the pull request:
right now `model.update('status', true)` is not working, as only the `model.status` field is updated, and not `model.attrs.status` (which is the field used in the `save()` method: `this._schema.db[collection].update(this.attrs.id, this.attrs);`).

samselikoff commented:
#719

samselikoff commented:
Thanks for kicking this off!

I don't think we need to modify the associations - they're taken care of, because associations must be defined on models. What we need to do is call [_definePlainAttribute](https://github.com/samselikoff/ember-cli-mirage/blob/master/addon/orm/model.js#L212) on attrs that are passed to `update` that the model wasn't aware of when it was instantiated.

I think we want to add a line [here](https://github.com/samselikoff/ember-cli-mirage/blob/master/addon/orm/model.js#L81) in the forEach that calls `_definePlainAttribute` for each attr. I believe `_definePlainAttribute` is idempotent (it can be called multiple times safely), but we should add a test for this.

Think you could update your PR?

Leooo commented:
I'm not sure how to use `_definePlainAttribute`, because I'm not sure what it does (define `get` / `set` on each attribute?, but then we still need to `set` the attribute later in the `update` hook, what's the difference with just writing `this.attrs[attr] = this[attr]` then??

The code on the belongsTo is just picked up from https://github.com/samselikoff/ember-cli-mirage/pull/648, just to insist this is really in need to be merged..

Leooo commented:
@samselikoff updated to use `_definePlainAttribute`, but we still need to add a line before`this.attrs[attr] = attrs[attr]` to make sure that `this.attrs.hasOwnProperty(attr))` is true. Also added a test. 

Removed the relationship part which was not part of this fix, but the PR in https://github.com/samselikoff/ember-cli-mirage/pull/568 seems very sensible to me (you intend to have a model returned by `server.create` but that's not the case for now and things need to work).

L

Leooo commented:
Also added the `configurable` option without which `Object.defineProperty` can not be called twice (and corresponding test which is failing without configurable).

samselikoff commented on addon/orm/model.js:
```diff
@@ -221,6 +223,7 @@ class Model {
 
     // Define the getter/setter
     Object.defineProperty(this, attr, {
+      configurable: true,
```
let's drop this and change line 215 from 

```js
if (this[attr] !== undefined) {
  return;
}
```

to

```js
if (this.hasOwnProperty(attr)) {
  return;
}
```

Leooo replied:
done, but wouldn't it be dangerous for users trying to set directly the property without using update? would mess things up then right?

Leooo replied:
Answer: no, I added a test for it

Leooo replied:
To make the test I added pass, it was necessary to remove the checks before Object.defineProperty, which means it will be run every time. Should be fine to do that I suppose.

samselikoff commented on bower.json:
```diff
@@ -6,6 +6,7 @@
     "ember-qunit-notifications": "0.1.0",
-    "sinonjs": "~1.17.1"
+    "sinonjs": "~1.17.1",
+    "Faker": "3.0.1"
```
don't need this, correct?

Leooo replied:
as you saw in https://github.com/samselikoff/ember-cli-mirage/issues/850 we do

Leooo replied:
correct

samselikoff commented on addon/orm/model.js:
```diff
@@ -79,6 +79,8 @@ class Model {
     }
 
     Object.keys(attrs).forEach(function(attr) {
+      this.attrs[attr] = attrs[attr];
```
I don't think we need this once you make the change below

Leooo replied:
true

samselikoff commented on tests/integration/orm/update-test.js:
```diff
@@ -71,3 +71,19 @@ test('it can update a hash of attrs immediately', function(assert) {
+  assert.deepEqual(this.db.users.find(1), { id: '1', name: 'Evil link', location: 'Hyrule', evil: true, reallyEvil: 'absolutely evil' });
+});
+
+test('it can override an existing attribute', function(assert) {
```
do we need this test? Seems similar to L67 above

Leooo replied:
not if you removed the `undefined` test above, indeed

samselikoff commented:
Awesome, thank you for your work, I think we're almost there!

samselikoff commented on addon/orm/model.js:
```diff
@@ -210,9 +211,6 @@ class Model {
    * @private
    */
   _definePlainAttribute(attr) {
-    if (this[attr] !== undefined) {
```
Instead change this to

```js
let existingProperty = Object.getOwnPropertyDescriptor(this, attr);
if (existingProperty && existingProperty.get) {
  return;
}
```

then we can remove `configurable: true`

Leooo replied:
@samselikoff done

samselikoff commented:
Starting to think we should have used a `.attrs` field on models. `link.someNewProp = 'foo'` is technically part of the api but this actually doesn't work:

```js
link.someKnownProp = 'foo';
link.someNewProp = 'bar';

link.save();
```

Save won't persist the unknown prop. 😕 

I guess the the best we can do for now is make `update` work with new attrs. Possible eventual paths forward are (1) tell users to interact with `.attrs` (2) use `.set` or (3) allow users to define a schema.

samselikoff commented:
One last comment then we can merge this. Thanks for all your work!

samselikoff commented:
🎊
